lib/AST/Module.cpp

```diff
@@ -1454,6 +1454,189 @@ const clang::Module *ModuleDecl::findUnderlyingClangModule() const {
   return nullptr;
 }
 
+//===----------------------------------------------------------------------===//
+// Cross-Import Overlays
+//===----------------------------------------------------------------------===//
+
+namespace swift {
+/// Represents a file containing information about cross-module overlays.
+class OverlayFile {
+  /// If not empty, contains the name of the cross-import file that must be
+  /// loaded; \c overlayModuleNames is not valid and must be empty.
+  ///
+  /// If empty, \c overlayModuleNames is valid and will contain the list of
+  /// module names unless loading the file failed.
+  StringRef filePath;
+
+  /// The list of module names; empty if loading failed.
+  llvm::TinyPtrVector<Identifier> overlayModuleNames;
+
+  /// Actually loads the overlay module name list. This should mutate
+  /// \c overlayModuleNames, but not \c filePath.
+  void loadOverlayModuleNames(ModuleDecl *M, SourceLoc diagLoc,
+                              Identifier bystandingModule);
+
+public:
+  // Only allocate in ASTContexts.
+  void *operator new(size_t bytes) = delete;
+  void *operator new(size_t bytes, const ASTContext &ctx,
+                     unsigned alignment = alignof(OverlayFile)) {
+    return ctx.Allocate(bytes, alignment);
+  }
+
+  OverlayFile(StringRef filePath)
+    : filePath(filePath), overlayModuleNames() { }
+
+  /// Returns the list of additional modules that should be imported if both
+  /// the primary and secondary modules have been imported. This may load a
+  /// file; if so, it will diagnose any errors itself and arrange for the file
+  /// to not be loaded again.
+  ///
+  /// The result can be empty, either because of an error or because the file
+  /// didn't contain any overlay module names.
+  ArrayRef<Identifier> getOverlayModuleNames(ModuleDecl *M, SourceLoc diagLoc,
+                                             Identifier bystandingModule) {
+    if (!filePath.empty()) {
+      assert(overlayModuleNames.empty()
+             && "cross-import list can't be full before loading");
+      loadOverlayModuleNames(M, diagLoc, bystandingModule);
+      filePath = "";
+    }
+    return overlayModuleNames;
+  }
+};
+}
+
+void ModuleDecl::addCrossImportOverlayFile(StringRef file) {
+  auto &ctx = getASTContext();
+
+  Identifier secondaryModule = ctx.getIdentifier(llvm::sys::path::stem(file));
+  declaredCrossImports[secondaryModule]
+      .push_back(new (ctx) OverlayFile(ctx.AllocateCopy(file)));
+}
+
+void ModuleDecl::
+findDeclaredCrossImportOverlays(Identifier bystanderName,
+                                SmallVectorImpl<Identifier> &overlayNames,
+                                SourceLoc diagLoc) {
+  if (getName() == bystanderName)
+    // We don't currently support self-cross-imports.
```

Comment:
This seems like this should be a hard `llvm::report_fatal_error` instead of a silent pass, in case someone tries to do this [same as with the empty case], although I don't see a reason for _why_ they would want to do this.

Reply:
`findDeclaredCrossImportOverlays()` is where other parts of the compiler can ask module X, "Do you have any cross-imports with module Y?" It is likely that, occasionally, callers will naïvely ask module X, "Do you have any cross-imports with module X?" If they do, emitting a fatal error seems kind of unfriendly—we should just quietly skip the check.

We could certainly also have a check for self-cross-imports (or self-overlaying) in the vicinity of `OverlayFile`, but that would be something different.

Reply:
That makes sense. Let me get back to this tomorrow.